src/main/drivers/rx_nrf24l01.c

```diff
@@ -0,0 +1,346 @@
+    // CE as OUTPUT
+    RCC_AHBPeriphClockCmd(NRF24_CE_GPIO_CLK_PERIPHERAL, ENABLE);
+
+    GPIO_InitStructure.GPIO_Pin = NRF24_CE_PIN;
+    GPIO_InitStructure.GPIO_Mode = GPIO_Mode_OUT;
+    GPIO_InitStructure.GPIO_Speed = GPIO_Speed_50MHz;
+    GPIO_InitStructure.GPIO_OType = GPIO_OType_PP;
+    GPIO_InitStructure.GPIO_PuPd = GPIO_PuPd_NOPULL;
+
+    GPIO_Init(NRF24_CE_GPIO, &GPIO_InitStructure);
+#endif
+
+#ifdef STM32F10X
+    // CS as output
+    RCC_APB2PeriphClockCmd(NRF24_CS_GPIO_CLK_PERIPHERAL, ENABLE);
+
+    gpio_config_t gpio;
+    gpio.mode = Mode_Out_PP;
+    gpio.pin = NRF24_CS_PIN;
+    gpio.speed = Speed_50MHz;
+
+    gpioInit(NRF24_CS_GPIO, &gpio);
+    // CE as output
+    RCC_APB2PeriphClockCmd(NRF24_CE_GPIO_CLK_PERIPHERAL, ENABLE);
+
+    gpio.mode = Mode_Out_PP;
+    gpio.pin = NRF24_CE_PIN;
+    gpio.speed = Speed_50MHz;
+
+    gpioInit(NRF24_CE_GPIO, &gpio);
+    // TODO: NRF24_IRQ as input
+#endif
+
+    DISABLE_NRF24();
+    NRF24_CE_LO();
+
+    spiSetDivisor(NRF24_SPI_INSTANCE, SPI_9MHZ_CLOCK_DIVIDER);
+#endif // UNIT_TEST
+    hardwareInitialised = true;
+}
+
+// standby configuration, used to simplify switching between RX, TX, and Standby modes
+static uint8_t standbyConfig;
+
+void NRF24L01_Initialize(uint8_t baseConfig)
+{
+    standbyConfig = BV(NRF24L01_00_CONFIG_PWR_UP) | baseConfig;
+    rf_setup = 0x0F;
+}
+
+#if !defined(UNIT_TEST)
+uint8_t NRF24L01_WriteReg(uint8_t reg, uint8_t data)
+{
+    ENABLE_NRF24();
+    spiTransferByte(NRF24_SPI_INSTANCE, W_REGISTER | (REGISTER_MASK & reg));
+    spiTransferByte(NRF24_SPI_INSTANCE, data);
+    DISABLE_NRF24();
+    return true;
+}
+
+uint8_t NRF24L01_ReadReg(uint8_t reg)
+{
+    ENABLE_NRF24();
+    spiTransferByte(NRF24_SPI_INSTANCE, R_REGISTER | (REGISTER_MASK & reg));
+    uint8_t data = spiTransferByte(NRF24_SPI_INSTANCE, 0xFF);
+    DISABLE_NRF24();
+    return data;
+}
+
+uint8_t NRF24L01_WriteRegisterMulti(uint8_t reg, const uint8_t data[], uint8_t length)
+{
+    ENABLE_NRF24();
+    uint8_t res = spiTransferByte(NRF24_SPI_INSTANCE, W_REGISTER | ( REGISTER_MASK & reg));
+    for (uint8_t i = 0; i < length; i++)
+    {
+        spiTransferByte(NRF24_SPI_INSTANCE, data[i]);
+    }
+    DISABLE_NRF24();
+    return res;
+}
+
+uint8_t NRF24L01_WritePayload(const uint8_t *data, uint8_t length)
+{
+    ENABLE_NRF24();
+    uint8_t res = spiTransferByte(NRF24_SPI_INSTANCE, W_TX_PAYLOAD);
+    for (uint8_t i = 0; i < length; i++)
+    {
+        spiTransferByte(NRF24_SPI_INSTANCE, data[i]);
+    }
+    DISABLE_NRF24();
+    return res;
+}
+
+uint8_t NRF24L01_ReadRegisterMulti(uint8_t reg, uint8_t data[], uint8_t length)
+{
+    ENABLE_NRF24();
+    uint8_t res = spiTransferByte(NRF24_SPI_INSTANCE, R_REGISTER | (REGISTER_MASK & reg));
+    for(uint8_t i = 0; i < length; i++)
+    {
+        data[i] = spiTransferByte(NRF24_SPI_INSTANCE, 0xFF);
+    }
+    DISABLE_NRF24();
+    return res;
+}
+
+uint8_t NRF24L01_ReadPayload(uint8_t *data, uint8_t length)
```

Comment:
IMO gcc is smart enough to optimize tail-call, resulting in single instruction overhead. 
It will also reduce code duplication, which is usually good think (and in will be easier to move this into SPI driver)